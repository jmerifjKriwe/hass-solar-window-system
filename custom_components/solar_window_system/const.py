"""Constants for Solar Window System integration."""

DOMAIN = "solar_window_system"

# Entity ID prefix for better organization and future extensions
ENTITY_PREFIX = "sws"
ENTITY_PREFIX_GLOBAL = "sws_global"

# Global Configuration Entity IDs
GLOBAL_CONFIG_ENTITIES = {
    # Configuration entities retained as entities
    "min_solar_radiation": {
        "platform": "input_number",
        "name": "Minimum Solar Radiation",
        "min": 0,
        "max": 1000,
        "step": 1,
        "default": 50,
        "unit": "W/m²",
        "icon": "mdi:white-balance-sunny",
        "category": "configuration",
    },
    "min_sun_elevation": {
        "platform": "input_number",
        "name": "Minimum Sun Elevation",
        "min": -20,
        "max": 90,
        "step": 1,
        "default": 10,
        "unit": "°",
        "icon": "mdi:weather-sunset-up",
        "category": "configuration",
    },
    "scenario_b_enable": {
        "platform": "input_boolean",
        "name": "Scenario B Enable",
        "default": False,
        "icon": "mdi:toggle-switch-off",
        "category": "configuration",
    },
    "scenario_c_enable": {
        "platform": "input_boolean",
        "name": "Scenario C Enable",
        "default": False,
        "icon": "mdi:toggle-switch-off",
        "category": "configuration",
    },
    # Operational controls
    "maintenance_mode": {
        "platform": "input_boolean",
        "name": "Maintenance Mode",
        "default": False,
        "icon": "mdi:tools",
        "category": "configuration",
    },
    "sensitivity": {
        "platform": "input_number",
        "name": "Sensitivity",
        "min": 0.5,
        "max": 2.0,
        "step": 0.1,
        "default": 1.0,
        "unit": "",
        "icon": "mdi:tune-variant",
        "category": "configuration",
    },
    # Debug entities
    "debug": {
        "platform": "input_text",
        "name": "Debug",
        "max": 255,
        "default": "",
        "icon": "mdi:bug-outline",
        "category": "debug",
    },
    # Sensor entities
    "window_with_shading": {
        "platform": "sensor",
        "name": "Window with Shading",
        "default": 0,
        "unit": "",
        "icon": "mdi:window-shutter",
        "category": "sensor",
    },
    "total_power": {
        "platform": "sensor",
        "name": "Total Power",
        "default": 0,
        "unit": "W",
        "icon": "mdi:lightning-bolt",
        "category": "sensor",
    },
    "total_power_direct": {
        "platform": "sensor",
        "name": "Total Power Direct",
        "default": 0,
        "unit": "W",
        "icon": "mdi:weather-sunny",
        "category": "sensor",
    },
    "total_power_diffuse": {
        "platform": "sensor",
        "name": "Total Power Diffuse",
        "default": 0,
        "unit": "W",
        "icon": "mdi:weather-partly-cloudy",
        "category": "sensor",
    },
    # Weather-related select entities
    "weather_warning_sensor": {
        "platform": "input_select",
        "name": "Weather Warning Sensor",
        "options": [],
        "default": "",
        "icon": "mdi:weather-cloudy-alert",
        "category": "configuration",
    },
    "weather_forecast_temperature_sensor": {
        "platform": "input_select",
        "name": "Weather Forecast Temperature Sensor",
        "options": [],
        "default": "",
        "icon": "mdi:thermometer",
        "category": "configuration",
    },
}
