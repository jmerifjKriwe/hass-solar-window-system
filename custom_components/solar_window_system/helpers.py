"""Shared helpers for Solar Window System custom component."""

from typing import Any

from homeassistant.helpers import entity_registry as er


async def get_temperature_sensor_entities(hass: Any) -> list[dict]:
    """
    Collect temperature sensor options for selectors.

    Returns a list of option dicts with keys:
      - value: the entity_id (what gets saved)
      - label: the friendly name to show in the UI

    Keeps filtering logic from before but exposes friendly names so the
    selector can display user-friendly labels while returning the entity_id
    when selected.
    """
    entity_registry = er.async_get(hass)
    options: list[dict] = []

    # Add an explicit "inherit" option displayed in the UI as "-1" value.
    # The label is provided via translation in the front-end; use a safe
    # fallback label here in case translations aren't loaded at call time.
    try:
        inherit_label = await hass.helpers.translation.async_gettext(
            "options.step.global_basic.data_description.option_inherit"
        )
    except (AttributeError, TypeError):
        # If translation helper isn't available in this context, use fallback.
        # TypeError can be raised if the translation key is not found.
        inherit_label = "Inherit (use parent value)"
    options.append({"value": "-1", "label": str(inherit_label)})

    for ent in entity_registry.entities.values():
        if (
            ent.entity_id.startswith("sensor.")
            and not ent.disabled_by
            and not ent.hidden_by
        ):
            state = hass.states.get(ent.entity_id)
            if state and state.attributes.get("unit_of_measurement") in (
                "°C",
                "°F",
                "K",
            ):
                label = (
                    state.name or state.attributes.get("friendly_name") or ent.entity_id
                )
                options.append({"value": ent.entity_id, "label": str(label)})

    return options
