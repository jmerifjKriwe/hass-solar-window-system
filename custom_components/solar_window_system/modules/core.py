"""
Core Solar Window Calculator functionality.

This module contains the main SolarWindowCalculator class and basic functionality.
"""

from __future__ import annotations

import logging
import time
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from homeassistant.core import HomeAssistant

    from .flow_integration import WindowCalculationResult

_LOGGER = logging.getLogger(__name__)


class SolarWindowCalculator:
    """
    Main calculator class for Solar Window System calculations.

    This class handles the core calculation logic and coordinates
    between different calculation modules.
    """

    def __init__(
        self,
        hass: HomeAssistant,
        defaults_config: dict[str, Any] | None = None,
        groups_config: dict[str, Any] | None = None,
        windows_config: dict[str, Any] | None = None,
    ) -> None:
        """Initialize the calculator."""
        self.hass = hass
        self.defaults_config = defaults_config or {}
        self.groups_config = groups_config or {}
        self.windows_config = windows_config or {}

        # Entity cache for performance
        self._entity_cache: dict[str, Any] = {}
        self._cache_timestamp: float | None = None
        self._cache_ttl = 30  # seconds

        # Global entry reference (set by coordinator)
        self.global_entry: Any = None

    def get_safe_state(self, entity_id: str, default: Any = 0) -> Any:
        """Get entity state safely with fallback."""
        try:
            state = self.hass.states.get(entity_id)
        except (AttributeError, KeyError, TypeError):
            return default
        else:
            return state.state if state else default

    def get_safe_attr(self, entity_id: str, attr: str, default: str | float = 0) -> Any:
        """Get entity attribute safely with fallback."""
        try:
            state = self.hass.states.get(entity_id)
            return getattr(state, attr, default) if state else default
        except (AttributeError, KeyError, TypeError):
            return default

    def _get_cached_entity_state(
        self, entity_id: str, default_value: Any | None = None
    ) -> Any:
        """Get cached entity state with TTL."""
        current_time = time.time()

        # Check if cache is still valid
        if (
            self._cache_timestamp is None
            or current_time - self._cache_timestamp > self._cache_ttl
        ):
            self._entity_cache.clear()
            self._cache_timestamp = current_time

        # Return cached value if available
        if entity_id in self._entity_cache:
            return self._entity_cache[entity_id]

        # Get fresh value and cache it
        value = self.get_safe_state(entity_id, default_value)
        self._entity_cache[entity_id] = value
        return value

    def _resolve_entity_state_with_fallback(
        self, entity_id: str, fallback: str, valid_states: set[str]
    ) -> str:
        """Resolve entity state with fallback logic."""
        if not entity_id:
            return fallback

        state = self._get_cached_entity_state(entity_id, fallback)
        return state if state in valid_states else fallback

    # Placeholder for methods that will be implemented in other modules
    def calculate_window_solar_power_with_shadow(
        self,
        effective_config: dict[str, Any],
        window_data: dict[str, Any],
        states: dict[str, Any],
    ) -> WindowCalculationResult:
        """Calculate solar power for a window including shadow effects."""
        # This will be implemented in calculations.py
        msg = "Implemented in calculations module"
        raise NotImplementedError(msg)

    def create_debug_data(self, window_id: str) -> dict[str, Any] | None:
        """Create comprehensive debug data for a specific window."""
        # This will be implemented in debug.py
        msg = "Implemented in debug module"
        raise NotImplementedError(msg)

    def calculate_all_windows_from_flows(self) -> dict[str, Any]:
        """Calculate all window shading requirements using flow-based configuration."""
        # This will be implemented in flow_integration.py
        msg = "Implemented in flow_integration module"
        raise NotImplementedError(msg)
