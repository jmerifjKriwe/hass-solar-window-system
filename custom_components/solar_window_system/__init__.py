"""
This Source Code Form is subject to the terms of the Mozilla Public
License, v. 2.0. If a copy of the MPL was not distributed with this
file, You can obtain one at https://mozilla.org/MPL/2.0/.
"""

"""Solar Window System integration package."""

import logging

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant, ServiceCall, callback
from homeassistant.helpers import device_registry as dr

from .const import DOMAIN
from .coordinator import SolarWindowSystemCoordinator

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    # Get update interval from global config
    update_interval_minutes = 1
    for config_entry in hass.config_entries.async_entries(DOMAIN):
        if config_entry.data.get("entry_type") == "global_config":
            update_interval_minutes = config_entry.options.get(
                "update_interval", config_entry.data.get("update_interval", 1)
            )
            break

    # Register recalculate service (only once)
    if not hass.services.has_service(DOMAIN, "recalculate"):

        async def handle_recalculate_service(call: ServiceCall) -> None:
            """Service to trigger recalculation for all or a specific window."""
            window_id = call.data.get("window_id")
            # Recalculate for all window_configs coordinators
            for config_entry in hass.config_entries.async_entries(DOMAIN):
                if config_entry.data.get("entry_type") == "window_configs":
                    coordinator = hass.data[DOMAIN][config_entry.entry_id][
                        "coordinator"
                    ]
                    await coordinator.async_refresh()
                    # If a specific window_id is given, optionally filter or log
                    if window_id:
                        # Optionally, you could implement per-window recalculation logic here
                        _LOGGER.info(f"Recalculation triggered for window: {window_id}")
                    else:
                        _LOGGER.info("Recalculation triggered for all windows.")

        hass.services.async_register(DOMAIN, "recalculate", handle_recalculate_service)
    """Set up Solar Window System from a config entry."""
    _LOGGER.debug("Setting up entry: %s", entry.title)
    _LOGGER.debug("Entry data: %s", entry.data)
    _LOGGER.debug("Entry ID: %s", entry.entry_id)

    device_registry = dr.async_get(hass)

    # Create device for global config
    if entry.title == "Solar Window System":
        _LOGGER.debug("Creating device for global config")
        device_registry.async_get_or_create(
            config_entry_id=entry.entry_id,
            identifiers={(DOMAIN, "global_config")},
            name="Solar Window System",
            manufacturer="SolarWindowSystem",
            model="GlobalConfig",
        )

        # Set up all platforms for this entry
        await hass.config_entries.async_forward_entry_setups(
            entry, ["sensor", "number", "text", "select", "switch"]
        )

    # Handle group configurations entry (subentry parent)
    elif entry.data.get("entry_type") == "group_configs":
        # Create devices for existing subentries
        await _create_subentry_devices(hass, entry)

        # Initialize coordinator for group configurations as well
        coordinator = SolarWindowSystemCoordinator(hass, entry, update_interval_minutes)

        # Store coordinator per entry to support multiple entries
        hass.data.setdefault(DOMAIN, {})
        hass.data[DOMAIN][entry.entry_id] = {"coordinator": coordinator}

        # Start coordinator updates
        await coordinator.async_config_entry_first_refresh()

        # Set up platforms for group configurations
        await hass.config_entries.async_forward_entry_setups(
            entry, ["select", "sensor"]
        )

        # Add update listener to handle new subentries
        entry.add_update_listener(_handle_config_entry_update)

    # Handle window configurations entry (subentry parent)
    elif entry.data.get("entry_type") == "window_configs":
        # Create devices for existing subentries
        await _create_subentry_devices(hass, entry)

        # Initialize coordinator for calculation updates
        coordinator = SolarWindowSystemCoordinator(hass, entry, update_interval_minutes)

        # Store coordinator in hass.data for access by binary sensors
        hass.data.setdefault(DOMAIN, {})
        hass.data[DOMAIN][entry.entry_id] = {"coordinator": coordinator}

        # Start coordinator updates
        await coordinator.async_config_entry_first_refresh()

        # Set up platforms for window configurations
        await hass.config_entries.async_forward_entry_setups(
            entry, ["select", "sensor", "binary_sensor", "number", "text", "switch"]
        )

        # Add update listener to handle new subentries
        entry.add_update_listener(_handle_config_entry_update)

    # Register service for manual device creation (only once)
    if not hass.services.has_service(DOMAIN, "create_subentry_devices"):

        async def create_subentry_devices_service(
            _call: ServiceCall,
        ) -> None:
            """Service to manually create subentry devices."""
            for config_entry in hass.config_entries.async_entries(DOMAIN):
                if config_entry.data.get("entry_type") in [
                    "group_configs",
                    "window_configs",
                ]:
                    await _create_subentry_devices(hass, config_entry)

        hass.services.async_register(
            DOMAIN, "create_subentry_devices", create_subentry_devices_service
        )

    _LOGGER.debug("Setup completed for: %s", entry.title)
    return True


@callback
async def _handle_config_entry_update(hass: HomeAssistant, entry: ConfigEntry) -> None:
    """Handle config entry updates to create devices for new subentries."""
    _LOGGER.debug("Config entry update detected for: %s", entry.title)
    if entry.data.get("entry_type") in ["group_configs", "window_configs"]:
        await _create_subentry_devices(hass, entry)

        # Reconfigure coordinator if this entry manages flows
        if DOMAIN in hass.data and entry.entry_id in hass.data[DOMAIN]:
            coordinator = hass.data[DOMAIN][entry.entry_id]["coordinator"]
            await coordinator.async_reconfigure()

        # Reload select platform to pick up new entities for new subentries
        _LOGGER.debug("Reloading platforms for updated entry: %s", entry.title)
        if entry.data.get("entry_type") in ("group_configs", "window_configs"):
            # Reload platforms to pick up entities for new subentries
            platforms = ["select", "sensor"]
            if entry.data.get("entry_type") == "window_configs":
                platforms.append("binary_sensor")
            await hass.config_entries.async_unload_platforms(entry, platforms)
            await hass.config_entries.async_forward_entry_setups(entry, platforms)


async def _create_subentry_devices(hass: HomeAssistant, entry: ConfigEntry) -> None:
    """Create devices for all subentries using proper Device Registry API."""
    device_registry = dr.async_get(hass)

    # The correct way to access subentries is through entry.subentries
    # This is a dict where keys are subentry IDs and values are ConfigSubentry objects

    if getattr(entry, "subentries", None):
        for subentry_id, subentry in entry.subentries.items():
            if subentry.subentry_type == "group":
                group_name = subentry.title
                _LOGGER.debug("Creating device for group subentry: %s", group_name)

                # This is the key: use config_subentry_id parameter
                # to link device to subentry
                device = device_registry.async_get_or_create(
                    config_entry_id=entry.entry_id,
                    config_subentry_id=subentry_id,  # Links device to subentry!
                    identifiers={(DOMAIN, f"group_{subentry_id}")},
                    name=group_name,
                    manufacturer="SolarWindowSystem",
                    model="GroupConfig",
                )
                _LOGGER.debug(
                    "Device created for subentry: %s (%s)", device.name, device.id
                )
            elif subentry.subentry_type == "window":
                window_name = subentry.title
                _LOGGER.debug("Creating device for window subentry: %s", window_name)

                # Create device for window subentry
                device = device_registry.async_get_or_create(
                    config_entry_id=entry.entry_id,
                    config_subentry_id=subentry_id,  # Links device to subentry!
                    identifiers={(DOMAIN, f"window_{subentry_id}")},
                    name=window_name,
                    manufacturer="SolarWindowSystem",
                    model="WindowConfig",
                )
                _LOGGER.debug(
                    "Device created for subentry: %s (%s)", device.name, device.id
                )
                _LOGGER.debug(
                    "Device config_entries_subentries: %s",
                    device.config_entries_subentries,
                )
            else:
                _LOGGER.debug(
                    "Skipping subentry (unknown type): %s", subentry.subentry_type
                )
    else:
        _LOGGER.debug("No subentries found in entry")


async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Unload a config entry."""
    _LOGGER.debug("Unloading entry: %s", entry.title)

    # Unload platforms if this is the global config entry
    if entry.title == "Solar Window System":
        return await hass.config_entries.async_unload_platforms(
            entry, ["sensor", "number", "text", "select", "switch"]
        )

    # Handle window_configs entry unloading
    if entry.data.get("entry_type") == "window_configs":
        # Clean up coordinator
        if DOMAIN in hass.data and entry.entry_id in hass.data[DOMAIN]:
            del hass.data[DOMAIN][entry.entry_id]

        return await hass.config_entries.async_unload_platforms(
            entry, ["select", "sensor", "binary_sensor"]
        )

    # Handle group_configs entry unloading
    if entry.data.get("entry_type") == "group_configs":
        # Clean up coordinator
        if DOMAIN in hass.data and entry.entry_id in hass.data[DOMAIN]:
            del hass.data[DOMAIN][entry.entry_id]
        return await hass.config_entries.async_unload_platforms(
            entry, ["select", "sensor"]
        )

    return True
