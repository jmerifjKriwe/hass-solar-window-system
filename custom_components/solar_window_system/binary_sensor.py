"""Binary sensor platform for Solar Window System (window subentries)."""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from homeassistant.components.binary_sensor import BinarySensorEntity
from homeassistant.helpers import device_registry as dr
from homeassistant.helpers import entity_registry as er
from homeassistant.helpers.restore_state import RestoreEntity
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .const import DOMAIN

if TYPE_CHECKING:  # pragma: no cover
    from homeassistant.config_entries import ConfigEntry
    from homeassistant.core import HomeAssistant
    from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback

    from .coordinator import SolarWindowSystemCoordinator

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(
    hass: HomeAssistant,
    entry: ConfigEntry,
    async_add_entities: AddConfigEntryEntitiesCallback,
) -> None:
    """Set up binary sensors for window subentries."""
    if entry.data.get("entry_type") != "window_configs":
        return

    # Get coordinator
    coordinator = None
    if DOMAIN in hass.data and entry.entry_id in hass.data[DOMAIN]:
        coordinator = hass.data[DOMAIN][entry.entry_id]["coordinator"]

    if not coordinator:
        _LOGGER.warning("No coordinator found for window_configs entry")
        return

    device_registry = dr.async_get(hass)

    if not entry.subentries:
        _LOGGER.warning("No window subentries found")
        return

    entities = []
    for subentry_id, subentry in entry.subentries.items():
        if subentry.subentry_type != "window":
            continue

        window_name = subentry.title
        window_device = None

        # Find device for this window subentry
        for device in device_registry.devices.values():
            device_has_entry = (
                device.config_entries and entry.entry_id in device.config_entries
            )
            subentries = device.config_entries_subentries
            device_has_subentry = (
                subentries
                and entry.entry_id in subentries
                and subentry_id in subentries[entry.entry_id]
            )
            if device_has_entry and device_has_subentry:
                for identifier in device.identifiers:
                    if (
                        identifier[0] == DOMAIN
                        and identifier[1] == f"window_{subentry_id}"
                    ):
                        window_device = device
                        break
                if window_device:
                    break

        if not window_device:
            _LOGGER.warning("Window device not found for: %s", window_name)
            continue

        entity = WindowShadingRequiredBinarySensor(
            coordinator, window_device, window_name
        )
        entities.append((entity, subentry_id))

    # Add entities with their subentry_ids
    for entity, subentry_id in entities:
        async_add_entities([entity], config_subentry_id=subentry_id)


class WindowShadingRequiredBinarySensor(
    CoordinatorEntity, BinarySensorEntity, RestoreEntity
):
    """Binary sensor indicating if shading is required for a window."""

    coordinator: SolarWindowSystemCoordinator

    def __init__(
        self,
        coordinator: SolarWindowSystemCoordinator,
        device: dr.DeviceEntry,
        window_name: str,
    ) -> None:
        """Initialize the shading_required binary sensor for a window."""
        super().__init__(coordinator)
        self._window_name = window_name
        window_slug = window_name.lower().replace(" ", "_").replace("-", "_")
        self._attr_unique_id = f"sws_window_{window_slug}_shading_required"
        self._attr_suggested_object_id = f"sws_window_{window_slug}_shading_required"
        # Original name with prefix for traceability; friendly name will be overridden
        self._attr_name = f"SWS_WINDOW {window_name} Shading Required"
        self._attr_has_entity_name = False
        self._attr_device_info = {
            "identifiers": device.identifiers,
            "name": device.name,
            "manufacturer": device.manufacturer,
            "model": device.model,
        }
        self._attr_icon = "mdi:shield-sun"
        self._friendly_label = "Shading Required"

    @property
    def is_on(self) -> bool:
        """Return True if shading is currently required."""
        if not self.coordinator.data:
            return False
        return self.coordinator.get_window_shading_status(self._window_name)

    @property
    def extra_state_attributes(self) -> dict[str, str | float | int | None]:
        """Return additional state attributes."""
        if not self.coordinator.data:
            return {}

        window_data = self.coordinator.get_window_data(self._window_name)
        if not window_data:
            return {}

        attributes = {}

        # Add calculation results as attributes
        if "solar_result" in window_data:
            solar_result = window_data["solar_result"]
            attributes.update(
                {
                    "solar_power": solar_result.get("power", 0),
                    "solar_power_direct": solar_result.get("power_direct", 0),
                    "solar_power_diffuse": solar_result.get("power_diffuse", 0),
                    "shadow_factor": solar_result.get("shadow_factor", 1.0),
                }
            )

        # Add shading decision details
        if "shade_reason" in window_data:
            attributes["shade_reason"] = window_data["shade_reason"]

        if "scenarios_triggered" in window_data:
            attributes["scenarios_triggered"] = ", ".join(
                window_data["scenarios_triggered"]
            )

        return attributes

    async def async_added_to_hass(self) -> None:
        """Restore previous state if available and set a clean friendly name."""
        await super().async_added_to_hass()
        # Restore previous state if available (for UI consistency, not logic)
        restored_state = await self.async_get_last_state()
        if restored_state is not None:
            self._restored_state = restored_state.state
        entity_registry = er.async_get(self.hass)
        if self.entity_id in entity_registry.entities:
            ent_reg_entry = entity_registry.entities[self.entity_id]
            new_friendly_name = self._friendly_label
            if ent_reg_entry.original_name != new_friendly_name:
                entity_registry.async_update_entity(
                    self.entity_id, name=new_friendly_name
                )
