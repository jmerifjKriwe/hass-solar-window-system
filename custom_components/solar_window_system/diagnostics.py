"""Diagnostics support for Solar Window System integration."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:  # pragma: no cover
    from homeassistant.config_entries import ConfigEntry  # pragma: no cover
    from homeassistant.core import HomeAssistant  # pragma: no cover


async def async_get_config_entry_diagnostics(
    _hass: HomeAssistant, entry: ConfigEntry
) -> dict[str, Any]:
    """Return diagnostics for a config entry."""
    # You can expand this to include more relevant diagnostics data
    return {
        "entry_data": entry.data,
        "data": {},
    }
