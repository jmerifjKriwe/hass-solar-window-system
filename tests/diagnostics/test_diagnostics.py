"""
Test diagnostics endpoint for Solar Window System integration.

This file contains a snapshot test for the config entry diagnostics output.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from pytest_homeassistant_custom_component.common import MockConfigEntry

from custom_components.solar_window_system.const import DOMAIN
from custom_components.solar_window_system.diagnostics import (
    async_get_config_entry_diagnostics,
)
from tests.helpers.test_framework import BaseTestCase

if TYPE_CHECKING:
    from syrupy.assertion import SnapshotAssertion

    from homeassistant.core import HomeAssistant


class TestDiagnostics(BaseTestCase):
    """Test diagnostics endpoint for Solar Window System integration."""

    test_type = "diagnostics"

    def get_required_fixtures(self) -> list[str]:
        """Return list of required fixture names for this test type."""
        return ["hass", "snapshot"]

    async def test_config_entry_snapshot(
        self, hass: HomeAssistant, snapshot: SnapshotAssertion
    ) -> None:
        """
        Snapshot test for config entry diagnostics output.

        This will create the initial snapshot when run with `--snapshot-update`.
        """
        entry = MockConfigEntry(
            domain=DOMAIN,
            title="Solar Window System",
            data={
                "entry_type": "global_config",
                "window_width": 1.5,
                "window_height": 2.0,
                "g_value": 0.6,
                "diffuse_factor": 0.2,
                "threshold_direct": 250,
                "threshold_diffuse": 120,
                "temperature_indoor_base": 22.0,
                "temperature_outdoor_base": 20.0,
            },
            source="user",
            entry_id="test_entry_id",
        )
        entry.add_to_hass(hass)

        result = await async_get_config_entry_diagnostics(hass, entry)
        # Use snapshot assertion for diagnostics output
        assert result == snapshot  # noqa: S101
