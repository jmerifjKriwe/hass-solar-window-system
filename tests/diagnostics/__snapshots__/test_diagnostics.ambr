# serializer version: 1
# name: TestDiagnostics.test_config_entry_snapshot
  dict({
    'data': dict({
    }),
    'entry_data': mappingproxy({
      'diffuse_factor': 0.2,
      'entry_type': 'global_config',
      'g_value': 0.6,
      'temperature_indoor_base': 22.0,
      'temperature_outdoor_base': 20.0,
      'threshold_diffuse': 120,
      'threshold_direct': 250,
      'window_height': 2.0,
      'window_width': 1.5,
    }),
  })
# ---
